draft-jmiller-json-proof-algorithms.md

```diff
@@ -363,6 +365,59 @@ The same JWP in compact serialization:
 ```
 
+
+## Salted HMAC
+
+The Salted HMAC (SH) algorithm is similar to Single Use (SU), but uses generated salts and HMAC instead of individual JWSs to protect each payload.  It requires less compute but can result in potentially larger presentation proof values.
+
+Like SU, it also does not support unlinkability if the same JWP is presented multiple times and requires an individually issued JWP for each presentation in order to fully protect privacy.
+
+### SU-based Setup
+
+This algorithm uses most of the same setup as the Single Use one, including the use of JWS as the signature mechanism and the holder providing a Presentation Key.
+
+The main difference with the SH algorithm is that instead of the issuer generating an Ephemeral Key for signing, it instead generates a Shared Secret using HMAC.  This Shared Secret is then used to generate unique salt values that become the input keys to HMAC the issuer protected header and each payload.
+
+### Issuer Protected Header
+
+The holder's Presentation Key JWK MUST be included in the issuer protected header using the `presentation_jwk` claim.  The issuer MUST validate that the holder has posession of this key through a trusted mechanism such as verifying the signature of a unique nonce value from the holder.
+
+The salt for the issuer header is generated using HMAC with the Shared Secret as the key and the value "issuer_header" as the value.  The issuer header JSON is then serialized using UTF-8 and encoded with base64url into an octet array.  The generated salt is used as the HMAC key with the octet array as the value, and the resulting hash value becomes the first input into the larger octet array that will be signed by the issuer.
+
+### Payloads
+
+A unique salt is generated for each payload using HMAC with the Shared Secret at the key and the values "payload_X" where "X" is replaced by the zero-based array index of the payload, for example "payload_0", "payload_1", etc.
+
+Each payload is then serialized using UTF-8 and encoded with base64url into an octet array.  The generated salt for that payload based on its index is used as the HMAC key with the encoded octet array as the value.  The resulting hash value is appended to the larger octet array that will be signed by the issuer.
```

Comment:
It's less invention, and since what we're actually doing for each payload is generating an authentication code I'd like to support any existing safe MAC.